copy/multiple.go

```diff
@@ -118,17 +190,23 @@ func (c *copier) copyMultipleImages(ctx context.Context) (copiedManifest []byte,
 	// Copy each image, or just the ones we want to copy, in turn.
 	instanceDigests := updatedList.Instances()
 	instanceEdits := []internalManifest.ListEdit{}
-	instanceCopyList := prepareInstanceCopies(instanceDigests, c.options)
+	instanceCopyList, err := prepareInstanceCopies(updatedList, instanceDigests, c.options)
+	if err != nil {
+		return nil, fmt.Errorf("preparing instances for copy: %w", err)
+	}
 	c.Printf("Copying %d of %d images in list\n", len(instanceCopyList), len(instanceDigests))
+	originalOptions := *c.options.DestinationCtx // THIS IS HACK and will not be merged till: PR#2050
 	for i, instance := range instanceCopyList {
+		copyOption := originalOptions          // THIS IS HACK and will not be merged till: PR#2050
+		c.options.DestinationCtx = &copyOption // THIS IS HACK and will not be merged till: PR#2050
 		// Update instances to be edited by their `ListOperation` and
 		// populate necessary fields.
 		switch instance.op {
 		case instanceCopyCopy:
 			logrus.Debugf("Copying instance %s (%d/%d)", instance.sourceDigest, i+1, len(instanceCopyList))
 			c.Printf("Copying image %s (%d/%d)\n", instance.sourceDigest, i+1, len(instanceCopyList))
 			unparsedInstance := image.UnparsedInstance(c.rawSource, &instanceCopyList[i].sourceDigest)
-			updated, err := c.copySingleImage(ctx, unparsedInstance, &instanceCopyList[i].sourceDigest, false)
+			updated, err := c.copySingleImage(ctx, unparsedInstance, &instanceCopyList[i].sourceDigest, true)
```

Comment:
This must be `false` at least in the `cannotModifyManifestListReason` reason cases, and I think it should be `false` in most other cases.

(Notably many callers will not configure any compression format at all, so what does setting `true` mean in that case?)

---

I have changed my mind a bit from that absolute position — I think it could make sense to be more strict if `EnsureCompressionVariantsExist` is set, that’s clearly an opt-in to a new behavior.

OTOH I’m not immediately sure what “more strict” means in corner cases: if the image already has a gzip-exclusive instance, creating a gzip-exclusive instance on the destination clearly makes sense. But if the source image has a mixed-algorithm layer, and because of that we don’t do a Zstd clone, creating what does setting `…match` mean? There is no single algorithm we can set `opt.compressionFormat` to.